src/helper/blob-tools/broad-brush-helper.js

```diff
@@ -171,23 +190,48 @@ class BroadBrushHelper {
 
         // Simplify before adding end cap so cap doesn't get warped
         this.simplify(1);
+        const handleVec = event.delta.normalize(options.brushSize / 2);
+        this.finalPath.add(new paper.Segment(
+            event.point.add(handleVec),
+            handleVec.rotate(90),
+            handleVec.rotate(-90)
+        ));
         this.finalPath.closePath();
 
-        // Add end cap
-        const circ = new paper.Path.Circle(event.point, options.brushSize / 2);
-        circ.fillColor = options.fillColor;
-        this.finalPath = this.union(this.finalPath, circ);
         // Resolve self-crossings
         const newPath =
             this.finalPath
                 .resolveCrossings()
                 .reorient(true /* nonZero */, true /* clockwise */)
                 .reduce({simplify: true});
-        newPath.copyAttributes(this.finalPath);
-        newPath.fillColor = this.finalPath.fillColor;
-        this.finalPath.remove();
-        this.finalPath = newPath;
-        this.steps = 0;
+        if (newPath !== this.finalPath) {
+            newPath.copyAttributes(this.finalPath);
+            newPath.fillColor = this.finalPath.fillColor;
+            this.finalPath.remove();
+            this.finalPath = newPath;
+        }
+        
+        // Try to merge end caps
+        for (const cap of this.endCaps) {
+            const temp = this.union(this.finalPath, cap);
+            if (temp.area >= this.finalPath.area &&
+                !(temp instanceof paper.CompoundPath && !(this.finalPath instanceof paper.CompoundPath))) {
+                this.finalPath = temp;
+            } else {
+                // If the union of the two shapes is smaller than the original shape,
+                // or it caused the path to become a compound path,
+                // then there must have been a glitch with paperjs's unite function.
+                // In this case, skip merging that segment. It's not great, but it's
+                // better than losing the whole path for instance. (Unfortunately, this
+                // happens reasonably often to scribbles, and this code doesn't catch
+                // all of the failures.)
+                this.finalPath.insertAbove(temp);
+                temp.remove();
+                log.warn('Skipping a merge.');
```

Comment:
Is this log desired in production? If it was just for testing, can it be removed?

Reply:
I thought it made sense to leave in for production. Unfortunately, it happens a lot. This will log when a union fails in paper.js (it doesn't catch all the failures though), which means that there is a bug that needs fixing somewhere. Right now those bad unions fail silently, and may be why when you draw a scribble the brush stroke suddenly disappears or inverts. Hopefully we can eventually work on debugging those bad unions in paperjs.